11 create fen function under board struct

abrahimt opened the pull request:
to_string and from_string are working now.

abrahimt committed:
Merge pull request #28 from chanson02/27-termion-cant-build-on-windows

27 termion cant build on windows

abrahimt committed:
FEN WORKS!

to_string and from_string working

chanson02 committed:
Merge branch '11-create-fen-function-under-board-struct' of github.com:abrahimt/KnO3 into 11-create-fen-function-under-board-struct

chanson02 committed:
Fixed merging issues

chanson02 commented on src/board.rs:
```diff
@@ -1,8 +1,9 @@
+use std::{io::stdout, u8};
 use crossterm::{
     execute,
-    style::{ Color, Print, ResetColor, SetBackgroundColor, SetForegroundColor},
```
Is this really proper rust? imo it's much nicer to have whitespace between brackets
```rust
{ Color, Print }
```

abrahimt replied:
Rust formatter doesn't agree

chanson02 commented on src/board.rs:
```diff
@@ -112,16 +111,28 @@ impl Chessboard {
 
     /// # Return: The color of the piece
     fn find_fg(&self, p: char) -> Color {
-        if p.is_uppercase() { Color::White } else { Color::Black }
```
Also this. Does this function really need to be 5 lines?

chanson02 commented on src/board.rs:
```diff
@@ -112,16 +111,28 @@ impl Chessboard {
     /// # Return: The color of the board at this position
     fn find_bkgnd(&self, rank: usize, file: usize) -> Color {
         if (rank + file) % 2 == 0 {
-            return Color::Rgb { r: 190, g: 140, b: 170 };
```
I know this is super nitty but fr fr on jaw brother

chanson02 commented on src/board.rs:
```diff
@@ -146,83 +157,264 @@ impl Chessboard {
-}
-//Below here is just ideas for functions:
+    pub fn from_string(&self, fen: &str) -> Chessboard {
+        let mut chessboard = Chessboard {
```
I think we can move this into it's own constructor. It could be nice in the future to have a `Chessboard::blank();` where it returns a new, empty board.

chanson02 commented on src/board.rs:
```diff
@@ -146,83 +157,264 @@ impl Chessboard {
+                } else {
+                    let square_index = 8 * (rank - 1) + file;
+                    // Update the corresponding bitboard based on the piece type and color
+                    if piece.is_ascii_lowercase() {
```
Why are we checking if piece is lowercase? Doesn't the match already do that?
We should be able to get rid of this nest.

abrahimt replied:
It's for speed

chanson02 replied:
Elaborate?

abrahimt replied:
Ahhh I see what you mean lol

chanson02 commented on src/board.rs:
```diff
@@ -146,83 +157,264 @@ impl Chessboard {
-    // }
-}
-//Below here is just ideas for functions:
+    pub fn from_string(&self, fen: &str) -> Chessboard {
```
We should make this into a constructor. Remove &self from the args.

chanson02 commented on src/board.rs:
```diff
@@ -146,83 +157,264 @@ impl Chessboard {
-// }
+        // Parse castling rights
+        let fen_castle = fen_parts[2];
+        let mut castles = 0;
```
This is a neat way to do this! How did you come up with such a clever solution?
Nit: replace `castles` with `chessboard.castling_rights` so we can reduce the line count by a little.
Also, why is it called castling_rights?

abrahimt replied:
I saw it called that in one of the FEN examples, I think. I can change it to something else if you want.

chanson02 commented on src/board.rs:
```diff
@@ -146,83 +157,264 @@ impl Chessboard {
+                let square_index = (rank - 1) * 8 + (file - 1);
+
+                // Determine the piece on the current square
+                let piece = if (self.white_pawns >> square_index) & 1 != 0 {
```
Surely, there is a way to put this into a for loop? I'm too tired to think of it tho

abrahimt replied:
Lmk when you think of something!

chanson02 replied:
https://github.com/chanson02/KnO3/commit/928dbc211faf96c28867efdab2bfcc20d763b8c2

chanson02 reviewed (changes requested):
Didn't get a chance to finish `to_string` but this should give you enough to work with for at least a bit.

abrahimt committed:
Review Updates

Fixed some things from the review

abrahimt committed:
Merge branch '11-create-fen-function-under-board-struct' of https://github.com/chanson02/KnO3 into 11-create-fen-function-under-board-struct

abrahimt committed:
Spelling mistakes

Fixed some of cooper's spelling mistakes!

abrahimt committed:
Fixing a nit

Reduced line count by a bit

abrahimt committed:
Cleanup pt.1

Just cleaning up

abrahimt committed:
Clean up pt.2

abrahimt committed:
Constructors

Added empty constructor and moved from_string to the top of impl

abrahimt committed:
Fixed up from_string

Reduced some lines and took out the self argument

abrahimt committed:
Fixed a tab

Had an extra tab when I copied from_string

abrahimt committed:
Formatted

Used auto formatter

chanson02 committed:
Let's test to see if this skips formatting

chanson02 committed:
Simplified logic by combining two match statements

chanson02 committed:
Rewrote to_string part of finding which pieces go where

chanson02 committed:
Cleaned up the look of the empty constructor

chanson02 committed:
Comments and spacing

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +42,117 @@ impl Chessboard {
+    /// * `fen` - The FEN to be converted to a Chessboard.
+    /// # Return: Chessboard with the position from the FEN.
+    pub fn from_string(fen: &str) -> Chessboard {
+        let mut chessboard = Chessboard::new();
```
We should probably use some function like `valid_fen() -> bool` and return an error if it's invalid

abrahimt replied:
I added valid_fen but didn't implement it. I'll make an issue for that.

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +42,117 @@ impl Chessboard {
+    /// Forsyth–Edwards Notation Parser
+    /// * `fen` - The FEN to be converted to a Chessboard.
+    /// # Return: Chessboard with the position from the FEN.
+    pub fn from_string(fen: &str) -> Chessboard {
```
We should break this function up into multiple parts.
at the very least
- `set_pieces(&mut self, FEN: &str)`
- `set_caslting_rights(&mut self, FEN: &str)`
- `set_en_passant(&mut self, FEN: &str)`
So we can reduce the length of this function and make it easier to test.

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +215,175 @@ impl Chessboard {
-    // }
-}
-//Below here is just ideas for functions:
+                for &(piece, mask) in self.get_pieces().iter() {
```
Instead of using a for loop here, how about `piece_at_position`?

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +215,175 @@ impl Chessboard {
+            string_array.push("- ".to_string());
+        } else {
+            // Convert en passant square to algebraic notation
+            let row = (self.en_passant - 1) / 8 + 1;
+            let col = (self.en_passant - 1) % 8;
+
+            let column_char = (b'A' + col) as char;
```
This is genius! Good work!

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +215,175 @@ impl Chessboard {
+        string_array.push(match self.castling_rights {
+            0 => "- ".to_string(),
+            rights => {
+                let mut rights_string = String::new();
 
-// generate_moves_for_piece(piece) {
-//     // Generate all possible moves for the given piece
-//     // Consider piece-specific movement rules (e.g., pawn's initial double move, castling for king, etc.)
-// }
+                // Check individual castling rights and append to rights_string
+                if rights & 0b1000 != 0 {
+                    rights_string.push('K');
+                }
+                if rights & 0b0100 != 0 {
+                    rights_string.push('Q');
+                }
+                if rights & 0b0010 != 0 {
+                    rights_string.push('k');
+                }
+                if rights & 0b0001 != 0 {
+                    rights_string.push('q');
+                }
+
+                rights_string.push(' ');
+                rights_string
+            }
+        });
```
instead of building a string for castling rights, you should set it equal to `rights_string = "KQkq"` and then `.filter()` to remove the unused ones. I think that would clean it up a lot.
If it comes up empty put in `-`

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +215,175 @@ impl Chessboard {
-//     // Switch turns: white_turn = !white_turn;
-// }
+        // Whose turn
+        string_array.push(if self.white_turn {
+            "w ".to_string()
+        } else {
+            "b ".to_string()
+        });
```
this can be minorly simplified
`if self.white_turn { "w " } else { "b "}.to_string()`

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +215,175 @@ impl Chessboard {
+    /// Forsyth–Edwards Notation Serializer
+    /// * `chessboard` - The chessboard position to be converted to a FEN.
+    /// # Return: FEN string representing the board's position.
+    pub fn to_string(&self) -> String {
```
This function is wayyyy too long, we should break it down into smaller pieces for testing.

chanson02 reviewed (changes requested):
Included my requests for to_string this time.

chanson02 committed:
Added a job for formatting

chanson02 committed:
Skip format on bkgnd color

chanson02 committed:
Formatted

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +42,117 @@ impl Chessboard {
+                    let square_index = 8 * (rank - 1) + file;
+                    // Update the corresponding bitboard based on the piece type and color
+                    match piece {
+                        'p' => chessboard.black_pawns |= pow(2, square_index),
+                        'r' => chessboard.black_rooks |= pow(2, square_index),
+                        'b' => chessboard.black_bishops |= pow(2, square_index),
+                        'k' => chessboard.black_king |= pow(2, square_index),
+                        'q' => chessboard.black_queen |= pow(2, square_index),
+                        'n' => chessboard.black_knights |= pow(2, square_index),
+                        'P' => chessboard.white_pawns |= pow(2, square_index),
+                        'R' => chessboard.white_rooks |= pow(2, square_index),
+                        'B' => chessboard.white_bishops |= pow(2, square_index),
+                        'K' => chessboard.white_king |= pow(2, square_index),
+                        'Q' => chessboard.white_queen |= pow(2, square_index),
+                        'N' => chessboard.white_knights |= pow(2, square_index),
```
There is a builtin [pow function](https://doc.rust-lang.org/std/?search=pow) in rust. That would be great if we could use it here instead of importing a new module.
This might need to become it's own issue, as we might need to rework what datatype we use for `file` and `rank`

abrahimt committed:
Removed imported pow module

Replaced imported pow module with rust in built pow

abrahimt committed:
Added fen_valid

It's empty so we have to implement it later.

abrahimt committed:
Split up to_string function

Broke it down to
set_pieces(&mut self, FEN: &str)
set_caslting_rights(&mut self, FEN: &str)
set_en_passant(&mut self, FEN: &str)

abrahimt committed:
Updates from review

Replaced for loop with piece_at_position. Made valid_fen always be true

abrahimt committed:
Updated according to review

Used .filter for catling_rights

abrahimt reviewed (commented):
The format build keeps failing.

chanson02 committed:
Fixed a typo

chanson02 commented on Cargo.toml:
```diff
@@ -7,3 +7,4 @@ edition = "2021"
 
 [dependencies]
 crossterm = "0.22"
+num-traits = "0.2"
```
remove unused import

chanson02 committed:
Removed unused import

chanson02 commented on .github/workflows/rust.yml:
```diff
@@ -10,10 +10,15 @@ env:
+  format:
     runs-on: ubuntu-latest
+    steps:
+      - uses: actions/chackout@v3
```
Fix typo

chanson02 committed:
Added some separation

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +220,175 @@ impl Chessboard {
         ]
     }
 
+    fn valid_fen(fen: &str) -> bool {
+        let is_valid: bool = true;
+        //Check if fen is valid
+        is_valid
+    }
```
Either fill this out or make an issue for it.

abrahimt replied:
There is an issue for it.

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +41,123 @@ impl Chessboard {
+        let mut chessboard = Chessboard::new();
+        if !Self::valid_fen(fen) {
+            println!("{} is an invalid FEN. Try again.", fen);
+            chessboard
```
We should return an error instead if there is a invalid FEN.

chanson02 replied:
Use the `Result` enum in rust.

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +41,123 @@ impl Chessboard {
+        if !Self::valid_fen(fen) {
+            println!("{} is an invalid FEN. Try again.", fen);
+            chessboard
+        } else {
```
Invert this nest so we can get rid of this indentation level

chanson02 committed:
give from_string the ability to crash

chanson02 committed:
Changed docstring

chanson02 committed:
Don't bother allocating space if we're just going to error anyway

chanson02 committed:
When building a board from a string, start with an empty board

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +41,123 @@ impl Chessboard {
+    /// * `fen` - The FEN to be converted to a Chessboard.
+    /// # Return: Chessboard with the position from the FEN.
+    pub fn from_string(fen: &str) -> Chessboard {
+        let mut chessboard = Chessboard::new();
```
If we start from a new board (i.e. a board setup for a new game) wouldn't that mess up adding pieces because there are already pieces on the board?

chanson02 committed:
Moved parsing FEN pieces into it's own function && made a FEN module

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +41,123 @@ impl Chessboard {
+                        file += 1; // Move to the next file
+                    }
+                }
+            }
```
We should put this into it's own function

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +41,123 @@ impl Chessboard {
+                    _ => 0b0,
+                };
+                chessboard.castling_rights |= v;
+            }
```
while we're at it, let's put this into a separate function too.

abrahimt replied:
Done.

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +41,123 @@ impl Chessboard {
+                        chessboard.en_passant = col_value + 8 * (row as u8 - 1);
+                    }
+                }
+            }
```
Give it it's own function.

abrahimt replied:
Done.

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +220,175 @@ impl Chessboard {
+                            != 0
+                    })
+                    .collect::<String>();
 
-// generate_legal_moves() {
-//     legal_moves = []
-
-//     for each piece in pieces_of_current_turn {
-//         possible_moves = generate_moves_for_piece(piece)
-//         for each move in possible_moves {
-//             if is_legal_move(move) {
-//                 legal_moves.push(move)
-//             }
-//         }
-//     }
-//
-//     return legal_moves
-// }
+                format!("{} ", rights_string)
+            }
+        });
+    }
 
-// generate_moves_for_piece(piece) {
-//     // Generate all possible moves for the given piece
-//     // Consider piece-specific movement rules (e.g., pawn's initial double move, castling for king, etc.)
-// }
+    fn set_en_passant(&self, string_array: &mut Vec<String>) {
+        if self.en_passant == 0 {
+            string_array.push("- ".to_string());
+        } else {
+            let row = (self.en_passant - 1) / 8 + 1;
+            let col = (self.en_passant - 1) % 8;
+            let column_char = (b'A' + col) as char;
+
+            string_array.push(format!("{}{}", column_char, row));
+            string_array.push(" ".to_owned());
+        }
+    }
+}
```
Break this down into more manageable functions as well.

chanson02 committed:
Added a format skip for place_pieces

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +45,93 @@ impl Chessboard {
+        }
+    }
+
+    //fn parse_positions
```
get rid of this. oops that was my bad.

abrahimt replied:
Get rid of what?

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +194,175 @@ impl Chessboard {
+                            != 0
+                    })
+                    .collect::<String>();
 
-// generate_legal_moves() {
-//     legal_moves = []
-
-//     for each piece in pieces_of_current_turn {
-//         possible_moves = generate_moves_for_piece(piece)
-//         for each move in possible_moves {
-//             if is_legal_move(move) {
-//                 legal_moves.push(move)
-//             }
-//         }
-//     }
-//
-//     return legal_moves
-// }
+                format!("{} ", rights_string)
+            }
+        });
+    }
 
-// generate_moves_for_piece(piece) {
-//     // Generate all possible moves for the given piece
-//     // Consider piece-specific movement rules (e.g., pawn's initial double move, castling for king, etc.)
-// }
+    fn set_en_passant(&self, string_array: &mut Vec<String>) {
+        if self.en_passant == 0 {
+            string_array.push("- ".to_string());
+        } else {
+            let row = (self.en_passant - 1) / 8 + 1;
+            let col = (self.en_passant - 1) % 8;
+            let column_char = (b'A' + col) as char;
+
+            string_array.push(format!("{}{}", column_char, row));
+            string_array.push(" ".to_owned());
+        }
+    }
+}
```
move these to the fen util module?

abrahimt replied:
Done.

chanson02 commented on src/fen_util.rs:
```diff
@@ -0,0 +1,33 @@
+use crate::board::Chessboard;
+
+#[rustfmt::skip]
+pub fn place_pieces(chessboard: &mut Chessboard, fen_rows: &str) {
```
Add a docstring to this function

abrahimt replied:
Done

chanson02 commented on src/fen_util.rs:
```diff
@@ -0,0 +1,33 @@
+use crate::board::Chessboard;
+
+#[rustfmt::skip]
+pub fn place_pieces(chessboard: &mut Chessboard, fen_rows: &str) {
```
Perhaps we should have this return a result so we can catch the errors downstream?
Ideally it would never error because we would check for a valid fen first, but what's ideal almost never happens.

abrahimt replied:
Yah I need to think more on what we should return.

chanson02 replied:
Let's make a separate issue so we can close this one.

abrahimt replied:
Issue created.

chanson02 committed:
More format skipping

abrahimt committed:
Updates according to review

Moved functions to fen_util. Added docstrings to fen_util functions.

abrahimt committed:
Updated according to review

Made changes according to the review

chanson02 commented on src/fen_util.rs:
```diff
@@ -0,0 +1,192 @@
+/// # Arguments
+/// 
+/// * `chessboard` - A mutable reference to the chessboard.
+/// * `fen_rows` - FEN string representing the piece placement on the board.
```
let's include an example string in the docs.

abrahimt replied:
Issue created.

chanson02 commented on src/fen_util.rs:
```diff
@@ -0,0 +1,192 @@
+/// 
+/// * `chessboard` - A mutable reference to the chessboard.
+/// * `string_array` - A mutable vector of strings to store intermediate FEN string components.
+pub fn set_pieces(chessboard: &Chessboard, string_array: &mut Vec<String>) {
```
`set_pieces` is a little too close to `place_pieces` for my liking.
Maybe `get_fen_placement`? And have it return a String rather than mutate one.
Mutating a string is going to make it much harder for testing and tracking down bugs.

abrahimt replied:
Issue created.

chanson02 commented on src/fen_util.rs:
```diff
@@ -0,0 +1,192 @@
+/// 
+/// * `chessboard` - A mutable reference to the chessboard.
+/// * `string_array` - A mutable vector of strings to store intermediate FEN string components.
+pub fn set_castling_rights(chessboard: &mut Chessboard, string_array: &mut Vec<String>) {
```
`get_fen_castles` and return String

abrahimt replied:
Issue created.

chanson02 commented on src/fen_util.rs:
```diff
@@ -0,0 +1,192 @@
+/// 
+/// * `chessboard` - A mutable reference to the chessboard.
+/// * `string_array` - A mutable vector of strings to store intermediate FEN string components.
+pub fn set_en_passant(chessboard: &mut Chessboard, string_array: &mut Vec<String>) {
```
`get_fen_passant` and return String

abrahimt replied:
Issue created.

chanson02 commented on src/fen_util.rs:
```diff
@@ -0,0 +1,192 @@
+pub fn parse_piece_placement(chessboard: &mut Chessboard, piece_placement: &str) -> Result<(), String> {
+    self::place_pieces(chessboard, piece_placement);
+    Ok(())
+}
```
Why handle this error here instead of directly in place_pieces?

abrahimt replied:
Yah I'll remove the error handling there.

chanson02 commented on src/board.rs:
```diff
@@ -45,13 +45,62 @@ impl Chessboard {
+
+    let mut chessboard = Chessboard::empty();
+
+    let fen_parts: Vec<&str> = fen.split_whitespace().collect();
```
You're going to hate me for this: Now that I'm thinking about it, this should always be of length 4 right? (positions, turn, castling, en_passant)

fen_parts should be an array rather than a vector, so we can use the stack instead of the heap. It's always better to be deliberate!

abrahimt replied:
It'll be length 6. We aren't using the last 2 parts yet.

abrahimt replied:
But yah I can try to figure something out for that.

abrahimt replied:
Issue created.

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +163,95 @@ impl Chessboard {
+    /// Forsyth–Edwards Notation Serializer
+    /// * `chessboard` - The chessboard position to be converted to a FEN.
+    /// # Return: FEN string representing the board's position.
+    pub fn to_string(&mut self) -> String {
```
Why is this taking a mutable self? I don't think any mutation should be happening in this function.

abrahimt replied:
Issue created.

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +163,95 @@ impl Chessboard {
+    /// * `chessboard` - The chessboard position to be converted to a FEN.
+    /// # Return: FEN string representing the board's position.
+    pub fn to_string(&mut self) -> String {
+        let mut string_array: Vec<String> = Vec::with_capacity(6);
```
Prefer array over vector here.

abrahimt replied:
Issue created.

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +163,95 @@ impl Chessboard {
-//     // For example, if the last move was a double-step pawn move, return the square where the opponent's pawn can capture en passant.
-// }
+        // Piece placement
+        fen_util::set_pieces(self, &mut string_array);
```
We should change this to
`string_array[0] = fen_util::set_pieces(self);`
Always better to be deliberate.

abrahimt replied:
Issue created.

chanson02 commented on src/board.rs:
```diff
@@ -106,123 +163,95 @@ impl Chessboard {
-//     // Similarly, maintain flags for black's castling rights
-// }
+        // Whose turn
+        string_array.push(if self.white_turn {
+            String::from("w ")
+        } else {
+            String::from("b ")
+        });
```
I think you can simplify this to
`if self.white_turn { "w " } else { "b " }.to_string()`
Making it a bit easier to read imo

abrahimt replied:
Done.

abrahimt committed:
Formatted the code

Used rustfmt

abrahimt committed:
Updated according to review

Changed names of some functions

chanson02 reviewed (approved):
Related Issues
https://github.com/chanson02/KnO3/issues/31
https://github.com/chanson02/KnO3/issues/34
https://github.com/chanson02/KnO3/issues/35
https://github.com/chanson02/KnO3/issues/36
https://github.com/chanson02/KnO3/issues/37
